### PR TITLE
fix(helm): Set namespace on templates that does not contain a namespace

### LINKED ISSUE #6058: Kustomize 5.8.0 no longer add namespaces to resources in helm charts without explicit namespace
### What happened?

Before kustomize 5.8.0 if i declare the namespace in the `/namespace` of the kustomization.yaml file it will add a namespace to all the resources without an explicit namespace.

After 5.8.0 it won't add the namespace, breaking several 3rd party charts where the resources doesn't have an explicit `namespace: {{ .Release.Namespace }}` template.

For example i can no longer install the mariadb-operator chart in argocd after the kustomize update because the resources no longer have a namespace.

As a workaround for this specific chart i am using the following:

```yaml
patches:
  - target:
      kind: (ConfigMap|ServiceAccount|Deployment|ServiceMonitor|Role|RoleBinding|Service)
    patch: |-
      - op: add
        path: /metadata/namespace
        value: mariadb-system
```

But this can break anytime the chart or values.yaml is updated.

### What did you expect to happen?

The resources must have a namespace like in previous versions.

### How can we reproduce it (as minimally and precisely as possible)?

```yaml
# kustomization.yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
namespace: mariadb-system
helmCharts:
  - name: mariadb-operator
    repo: https://helm.mariadb.com/mariadb-operator
    version: 25.10.4
    releaseName: mariadb-operator
    namespace: mariadb-system
```

### Expected output

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app.kubernetes.io/instance: mariadb-operator
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: mariadb-operator
    app.kubernetes.io/version: 25.10.4
    helm.sh/chart: mariadb-operator-25.10.4
  name: mariadb-operator
  namespace: mariadb-system
spec:
...
```

### Actual output

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app.kubernetes.io/instance: mariadb-operator
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: mariadb-operator
    app.kubernetes.io/version: 25.10.4
    helm.sh/chart: mariadb-operator-25.10.4
  name: mariadb-operator
spec:
...
```

### Kustomize version

5.8.0

### Operating system

Linux

### LINKED ISSUE #6077: kustomize 5.8.1 references generated configMaps and secrets incorrectly in Helm generated deployments
### What happened?

kustomize v5.8.1 does not use the correct name of a generated configMap or secret when patching a deployment generated by Helm. It just uses the name without the hash. The same works flawlessly with kustomize  v5.7.1
Please see the attached reproducer example.

[kustomize-helm-configMap-reproducer.tgz](https://github.com/user-attachments/files/25263472/kustomize-helm-configMap-reproducer.tgz)

### What did you expect to happen?

Same behaviour as in kustomize v5.7.1 and earlier.

### How can we reproduce it (as minimally and precisely as possible)?

see attached files

### Expected output

```yaml
apiVersion: v1
data:
  config: foo
kind: ConfigMap
metadata:
  name: test-config-g46hh6k8tf
--
        secret:
          defaultMode: 292
          secretName: test-secret-gh24bh7t8g
      - configMap:
          defaultMode: 292
          name: test-config-g46hh6k8tf
```

### Actual output

```yaml
apiVersion: v1
data:
  config: foo
kind: ConfigMap
metadata:
  name: test-config-g46hh6k8tf
--
        secret:
          defaultMode: 292
          secretName: test-secret
      - configMap:
          defaultMode: 292
          name: test-config
```

### Kustomize version

v5.8.1

### Operating system

MacOS

### PR DESCRIPTION
**Pull request by Skaronator**
This PR ensures a namespace is set on Helm-generated resources only if the field is completely missing. This fixes #6058 and fixes #6077 and provides the best of both worlds: Helm charts with pre-defined namespaces remain untouched, while charts lacking a namespace field now function as expected.

---

The namespace is now set in the following priority:

1. Value from helm manifest (for charts already containing a namespace field)
2. Value from `helmChart[].namespace` (if set)
3. Value from `namespace` kustomization.yaml (If Set)

We don't modify any Namespaces relevant fields outside the metadata block. Helm Charts should already take care of that. (e.g. spec field for ClusterRoleBinding)

---

Checked the issue in #6058 and it is fixed.
```
/home/niklas/go/bin/kustomize build --enable-helm | grep -A 10 Deployment
kind: Deployment
metadata:
  labels:
    app.kubernetes.io/instance: mariadb-operator
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: mariadb-operator
    app.kubernetes.io/version: 25.10.4
    helm.sh/chart: mariadb-operator-25.10.4
  name: mariadb-operator
  namespace: mariadb-system
spec:
```

Checked the issue #6077 and it is fixed.

```
/home/niklas/go/bin/kustomize build --enable-helm | grep -C 5 gh24bh7t8g
apiVersion: v1
data:
  secret: Zm9v
kind: Secret
metadata:
  name: test-secret-gh24bh7t8g
  namespace: cluster
type: Opaque
---
apiVersion: v1
kind: Service
--
      serviceAccountName: release-name-test-chart
      volumes:
      - name: test-secret
        secret:
          defaultMode: 292
          secretName: test-secret-gh24bh7t8g
      - configMap:
          defaultMode: 292
          name: test-config-g46hh6k8tf
        name: test-conf
---
```

### CODE REVIEW

File: examples/helmNamespace/README.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  --helm-command "$helmCommand" \
+  "$EXAMPLE_HOME")
+printf '%s\n' "$output"
+```
+
+The Service is emitted by the chart without a namespace, so the example namespace should be applied:
+
+<!-- @checkMissingNamespaceFilled @testHelm -->
+```sh
+printf '%s\n' "$output" | grep -A4 'name: chart-service' | grep 'namespace: top-level-ns'
```
Comment (Skaronator) [inline examples/helmNamespace/README.md]: Only thing I'm unsure is this result here.

When the namespace its completely missing kustomize will use the top level `namespace` field in the kustomize file and not the value in `helmCharts.namespace`.

I think for consistency it would be better if helmCharts.namespace can overwrite top level `namespace`. Which is what happens when helm charts actually use `{{ .Release.Namespace }}`.

Comment (Skaronator) [reply examples/helmNamespace/README.md]: My last commit fixes this issue and apply the namespace consistently.

**Review by mfredenhagen — APPROVED** [review]
Built from source and tested manually 👍

**Review by davinkevin — APPROVED** [review]

**Comment by luggesexe** [comment]
What is the current state here? When could this be released? I encounter the same mariadb issue referenced.

**Comment by luschmgit** [comment]
Good idea! Ran into the same issue while migrating from ArgoCD 3.0.12 (kustomize 5.6.0) to 3.4.1 (kustomize 5.8.1). This feature would have saved me a couple of hours investigating and having to set the namespace explicitly for each resource in the template.

**Comment by koba1t** [comment]
@Skaronator 

Thank you for addressing this issue!

Does this PR implementation determine whether to apply the namespace transformer to resources within Helm charts based on whether a namespace is specified in the original helmCharts?

I believe this implementation can resolve the scenarios described in the two currently open issues without changing the API. However, I don't think it covers cases where the kustomize functionality prior to v5.7.0 was used to forcibly overwrite a namespace that was already set in helmCharts. (For example, if someone want to use a chart containing resources with the `kube-system` namespace in an environment where `kube-system` is unavailable due to rbac permission.)

Therefore, I think it would be better to add a new field like `forceOverwriteNamespace: True` to HelmCharts and restore the old behavior if that flag exists. https://github.com/kubernetes-sigs/kustomize/issues/6058#issuecomment-3998096168

If you have a different opinion, please let me know!

**Comment by Skaronator** [comment]
@koba1t

> Does this PR implementation determine whether to apply the namespace transformer to resources within Helm charts based on whether a namespace is specified in the original helmCharts?

The logic from my original PR (which caused the initial breakage, https://github.com/kubernetes-sigs/kustomize/pull/5940) and your follow-up PR (https://github.com/kubernetes-sigs/kustomize/pull/6044) remains unchanged.

This means the `namespace` field passed to Helm is evaluated in the following priority (from highest to lowest):

* `namespace` set directly within the `helmCharts` block.
* `namespace` set within the same `kustomization.yaml` file.
* `namespace` inherited from a parent `kustomization.yaml` file.


> Therefore, I think it would be better to add a new field like `forceOverwriteNamespace: true` to HelmCharts and restore the old behavior if that flag exists.

While a flag like `forceOverwriteNamespace: true` might seem to cover all use cases, I don't think it is the right path forward.

The current PR implementation tells `helm` the correct namespace up front, and if `metadata.namespace` is missing from the rendered output, we simply set it. Crucially, we avoid touching fields like `ClusterRoleBinding.subjects.namespace` (which the `NamespaceTransformer` normally modifies) because well-maintained Helm charts should manage those fields themselves.

Take the Grafana Alloy Helm Chart as an example:

```yaml
subjects:
  - kind: ServiceAccount
    name: {{ include "alloy.serviceAccountName" . }}
    namespace: {{ include "alloy.namespace" . }}
```
[Source](https://github.com/grafana/alloy/blob/02883fbf282609693b504f206eb50033aedc9053/operations/helm/charts/alloy/templates/rbac.yaml#L68)

In my opinion, introducing `forceOverwriteNamespace: true` isn't a robust solution. Charts designed to be deployed across multiple namespaces typically make those namespaces configurable via [values.yaml](url). It is almost always better to configure namespace settings natively through `values.yaml` rather than trying to force modifications externally after the chart is rendered, which is often incomplete or prone to breaking.

Another great example is cert-manager, which lets you configure `leaderElection.namespace`:

```yaml
  leaderElection:
    # Override the namespace used for the leader election lease.
    namespace: "kube-system"
```
[Source](https://github.com/cert-manager/cert-manager/blob/1ea4e0ccf03d8a8a148b4efc0885967703e64b68/deploy/charts/cert-manager/values.yaml#L75)

This value is used in RBAC resources, which the `NamespaceTransformer` _could_ technically change. However, the chart also embeds this exact value directly into the deployment's container arguments:

```yaml
          {{- with .Values.global.leaderElection }}
          - --leader-election-namespace={{ .namespace }}
          {{- if .leaseDuration }}
```
[Source](https://github.com/cert-manager/cert-manager/blob/1ea4e0ccf03d8a8a148b4efc0885967703e64b68/deploy/charts/cert-manager/templates/cainjector-deployment.yaml#L89-L91)

If we relied on `forceOverwriteNamespace: true`, the cert-manager deployment would break. The `NamespaceTransformer` would forcefully update the RBAC resources, but the `--leader-election-namespace` container argument would remain completely untouched, resulting in a non-functioning deployment.

All things considered, I'd say this PR fixes 99.5% of the issues people experienced with Kustomize 5.8.0 and 5.8.1 (and again, sorry about that!).

**Comment by koba1t** [comment]
@Skaronator 

I understand your approach to implementing this PR.

Initially, I was considering restoring the functionality of the previous `helmCharts` for compatibility reasons, but I agree with the suggestion to align it with the behavior of `helm template | kubectl apply -f -` as in this PR.

Therefore, I would like to proceed with this implementation approach.

**Comment by koba1t** [comment]
@Skaronator 
The PR description says we don’t modify namespace-relevant fields outside metadata, but the current NamespaceTransformer can still populate subjects[].namespace for default ServiceAccount subjects.
https://github.com/kubernetes-sigs/kustomize/blob/313aacedd3adcb6564fd362bcef52f5aa3abef85/api/filters/namespace/namespace.go#L31
I think Helm-generated resources should either force SetRoleBindingSubjects: none or use a metadata-only namespace fill path.

Could you consider to avoid applying the RoleBinding/ClusterRoleBinding subject namespace logic to Helm-generated resources?

**Comment by Skaronator** [comment]
@koba1t Good catch, you're right. Even though the namespace fill is metadata-only, the transformer was still running the RoleBinding/ClusterRoleBinding subject logic on Helm-generated resources, which could populate `subjects[].namespace` for `default` ServiceAccount subjects.

I went with your first suggestion: for Helm-generated resources the `NamespaceTransformer` now forces `SetRoleBindingSubjects: none`, so subject namespaces are left entirely to the chart. The metadata-only namespace fill behavior is unchanged.

I also added a regression test `TestNamespaceTransformer_HelmOriginDoesNotSetRoleBindingSubjects` that confirms a Helm-generated `RoleBinding` with a `default` subject gets its `metadata.namespace` filled while `subjects[].namespace` is left untouched.

**Comment by koba1t** [comment]
Thanks for your work!

To prevent regression, if possible, could you also add a test to confirm that #6077 has been fixed?

**Comment by Skaronator** [comment]
@koba1t I added both an unit test and an integration test.

**Comment by koba1t** [comment]
@Skaronator 
Thanks for your fix!

Sorry, there is one point I was wondering about.
It appears that the Helm namespace is currently being handled by a namespace transformer; will this still work if the `namespace` is not specified in `kustomization.yaml`?

Example
```
## no namespace transformer
helmCharts:
- name: example
  namespace: helm-ns
```

Could you please add a test to verify that it works correctly in such cases?

**Comment by davidandreoletti** [comment]
@koba1t Any idea when this would be released ? 

I have a OpenEBS deployment whose [Chart.yaml includes several dependencies](https://github.com/openebs/openebs/blob/e5353bdb582dcb59defa4f8a74fbef2150205ef8/charts/Chart.yaml#L17) such as Mayastor chart. With kubectl's kustomize v5.8.1, the mayastor chart namespace is **never set** and result in mayastor resources deployed in the 'default' namespace while they should be in a namespace defined by helmCharts.namespace OR the namespace transformer.

**Comment by Skaronator** [comment]
I did a review with Claude Fable 5 and it found some blind spots which are now solved. I also added a few more integration tests to hopefully cover every edge case.

**Review by saitejabandaru-in — APPROVED** [review]
LGTM! Excellent architectural fix for preserving chart-defined namespaces while still populating missing namespaces from `helmCharts[].namespace`. Resolves the hash dropping issue elegantly.

**Comment by koba1t** [comment]
@Skaronator 

Could you clarify whether the change to `SetRoleBindingSubjects` in `ecad3ec0` was intentional?
I think that commit was undo a change from below review.
https://github.com/kubernetes-sigs/kustomize/pull/6108#issuecomment-4699613631

**Review by chengfang — APPROVED** [review]

**Comment by koba1t** [comment]
@Skaronator 
Would you take time to work this?
If possible, I would like to include a fix for this issue in the next release.
